### ISSUE TITLE
Feature request/idea Complete delegation of email templates 

### ISSUE THREAD

**Issue by joshchernoff** [pow-auth/pow#469]
I'd like to have bamboo deal 100% with rendering and delivery of emails for my app. 

You can see a work in progress here where I delegate my pow mailer here:
https://github.com/PolymorphicProductions/morphic.pro/blob/bamboo_mailer/lib/morphic_pro_web/pow/mailer.ex

To my bamboo mailer here:
https://github.com/PolymorphicProductions/morphic.pro/blob/bamboo_mailer/lib/morphic_pro_web/email.ex

I personally think email is not exclusive to pow and that it's better to have a single consolidated module that deals with the mail needs holistically. For me, that's bamboo as I know it best for my needs. 

The current issue I have with pows extensions is that even if I delegate the mailer to bamboo, by the time the email has been created by pow its already rendered the HTML templates to its html prop which is not needed as I will be overriding it anyways via bamboo once I get to pass that data to it. I'd also like to avoid having to make a view for each extension as well. 

IE the email that gets passed to the miles cast function as already rendered the view in the html. This is what I want to delegate to bamboo. 

Secondly, the next issue I have currently is that I need access to the conn in my view's render which if I'm delegating from cast/1 I could gain access to it from the email but sadly `Pow.Phoenix.Mailer.Mail.new/4` does not pass its conn alone with the struct it makes. Whats your thoughts about that? 

Its also in `Pow.Phoenix.Mailer.Mail.new/4` where I see the views render is called for the html and text values and I was trying to think how I could configure this mail to skip that altogether if I want to have that delegated to something else later on. 

Anyhow thanks for all the great work thus far.

**Comment by danschultzer** [comment]
So there are two concerns here:

1. Render email templates and subject
2. Deliver the email

And the way I handle it is that the controller does both, as seen in the [email confirmation callbacks](https://github.com/danschultzer/pow/blob/v1.0.19/lib/extensions/email_confirmation/phoenix/controllers/controller_callbacks.ex#L121-L134):

```elixir
  @doc """
  Sends a confirmation e-mail to the user.
  The user struct passed to the mailer will have the `:email` set to the
  `:unconfirmed_email` value if `:unconfirmed_email` is set.
  """
  @spec send_confirmation_email(map(), Conn.t()) :: any()
  def send_confirmation_email(user, conn) do
    url               = confirmation_url(conn, user)
    unconfirmed_user  = %{user | email: user.unconfirmed_email || user.email}
    email             = Mailer.email_confirmation(conn, unconfirmed_user, url)

    Pow.Phoenix.Mailer.deliver(conn, email)
  end
```

My thinking was that this is how Phoenix handles render and return response. However, after going through the Phoenix code base again , I see that the [the controller actually does both at the same time](https://github.com/phoenixframework/phoenix/blob/06565eff598f117a879ea202d42015fdacb6dcf2/lib/phoenix/controller.ex#L781-L793). So I think you're right that the current way Pow does this doesn't make sense.

Ideally a `Mail` struct is set up, then `cast/1` is called, and then before `process/1` is called the render happens. This way you can override the behavior if necessary. I'll have to think a bit more about this though, been busy these days.

**Comment by joshchernoff** [comment]
For me ideally, the concerns are as followed. 

Given I'm able to configure this workflow that state I want to delegate the following to bamboo from pow:

Pow:
1: creates the shape of the email data given its not been gated by logic. IE: sender, receiver, subject, attrs and conn. 
2: Pow calls cast, which I use do delegate to bamboo the needed data from pow. 

Bamboo:
3: creates an email, hydrates the emails layouts, templates, headers ect. 
4: At this point, you could send the email back to pow so that it could be sent off to process which could then would again delegate to bamboo for deliver_now/deliver_later

In short, the only thing I really want pow to dow\ with my emails is the control flow up to the point its time to make an email stuct. From there out I'm trying to hand all the remaining work off to bamboo.

**Comment by joshchernoff** [comment]
Thoughts about the `send_confirmation_email/2` in the controller callbacks example. Is the idea to possibly delegate the email and deliver calls to an intermediate module that would know based on config which modules to delegate the calls too?

**Comment by danschultzer** [comment]
> Is the idea to possibly delegate the email and deliver calls to an intermediate module that would know based on config which modules to delegate the calls too?

Could be, though I think the approach I detailed above would solve all your issues. In short it would look like this:

```elixir
mailer = # your mail module
type = # the unique mail name

type
|> Mail.new(assigns)
|> mailer.cast()
|> Mail.render() <--- this can be stopped by the above transformation
|> mailer.deliver()
```

Also, I haven't completely fleshed out the idea, but ideally you won't need any knowledge of the `conn` to build and deliver the email. This is what you would need for background jobs and such. It would be nice if you could easily call the `send_confirmation_email/2` email from any other part of the system, right now it sits in the `PowEmailConfirmation.Phoenix.ControllerCallbacks` and I don't think it belongs there.

**Comment by joshchernoff** [comment]
Have you looked at protocols yet? https://elixir-lang.org/getting-started/protocols.html
I'm thinking this could be a way to abstract away the implementation of each kind of email ("mail") into their own respective extension.

**Comment by danschultzer** [comment]
Yeah, I really like that concept, but I don't think it would fit for this situation?

**Comment by joshchernoff** [comment]
Sorry for being vague, the general idea I was kicking around in my head was. 
First: I think it would be best to try and remove the construct of an email from in the main pow lib. 
IE everything from `pow/lib/pow/phoenix/mailer/` should be migrated into its own lib/extension as pow its self should have to know about an email since it does not deal directly with them, only its extensions do. 

From there I was thinking about how to define a behavior/protocol to define what an email should look/act like which each extension could implement on their own respectively. 

In the end I was really just trying to think of better ways to decouple emails from the main lib as that's really what my main goal was really about in the first place. In that want to 100% delegate all that to bamboo.

**Comment by danschultzer** [comment]
> First: I think it would be best to try and remove the construct of an email from in the main pow lib.
> IE everything from `pow/lib/pow/phoenix/mailer/` should be migrated into its own lib/extension as pow its self should not have to know about an email since it does not deal directly with them, only its extensions do.

Yup 💯. It would remove an unnecessary part in core. I think it should be moved over the the extensions utilities though.

> From there I was thinking about how to define a behavior/protocol to define what an email should look/act like which each extension could implement on their own respectively.
> 
> In the end I was really just trying to think of better ways to decouple emails from the main lib as that's really what my main goal was really about in the first place. In that want to 100% delegate all that to bamboo.

Hmm yeah, though it would be possible to do this with the change in call logic I got in mind. Imagine this setup:

```elixir
def pow_email(%Mail{type: :pow_email_confirmation_confirm_email, user: user} = mail) do
  base_email()
  |> to(user.email)
  |> subject("Your custom subject")
  |> render("confirm_email.html")
end

# ...
```

It's triggered when `cast/2` is called. The fallback logic would be the current render handling.

Though maybe what you are thinking about is actually replacing the whole mailer call, so you can ignore the cast/process handling in the Pow mailer. A mail render call could look like this:

```elixir
my_mailer.pow_email_confirmation_send_confirmation_email(user_or_assigns)
```

And you can impl the mails with a macro:

```elixir
use Pow.Extension.Mailer,
  extensions: [...]

def pow_email_confirmation_send_confirmation_email(...) do
  base_email()
  |> to(user.email)
  |> subject("Your custom subject")
  |> render("confirm_email.html")
  |> deliver_now()
end
```

Also I would most likely drop `conn` complete, since this call can also come from another place than the endpoint/plug/controller. I think these are some great thoughts, I'll try work it through and see in what way this can be implemented.

**Comment by joshchernoff** [comment]
Awesome, I like where this is going. Thank you for entertaining my ideas. 🙏Aswell great work on this lib. I'm really enjoying it.
